File: backend/app/api/v1/endpoints/moderator.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.api.deps import get_db, get_current_user
from app.models.user import User
from pydantic import BaseModel

router = APIRouter()

# DTO-k
class RoleUpdate(BaseModel):
    role_id: int

class StatusUpdate(BaseModel):
    disabled: bool

# Csak moderátorok és adminok érhetik el
def require_moderator(current_user: User = Depends(get_current_user)):
    if current_user.role_id < 1:  # 1: moderator, 2: admin
        raise HTTPException(status_code=403, detail="Nincs jogosultságod ehhez a művelethez")
    return current_user

# Összes felhasználó listázása
@router.get("/users")
def list_all_users(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_moderator)
):
    users = db.query(User).all()
    return [
        {
            "id": u.id,
            "username": u.username,
            "email": u.email,
            "full_name": u.full_name,
            "phone": u.phone,
            "role_id": u.role_id,
            "created_at": u.created_at.isoformat(),
            "updated_at": u.updated_at.isoformat(),
        }
        for u in users
    ]

# Felhasználó szerepkör módosítása
@router.patch("/users/{user_id}/role", status_code=204)
def update_user_role(
    user_id: int,
    payload: RoleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_moderator)
):
    user = db.get(User, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="Felhasználó nem található")
    
    # Ne lehessen saját magad szerepkörét módosítani
    if user.id == current_user.id:
        raise HTTPException(status_code=400, detail="Nem módosíthatod a saját szerepkörödet")
    
    # Csak admin módosíthat admin jogot (adhat vagy elvehet)
    if payload.role_id >= 2 and current_user.role_id < 2:
        raise HTTPException(status_code=403, detail="Csak admin adhat admin jogosultságot")
    
    # Csak admin vehet el admin jogot
    if user.role_id >= 2 and current_user.role_id < 2:
        raise HTTPException(status_code=403, detail="Csak admin vehet el admin jogosultságot")
    
    user.role_id = payload.role_id
    db.add(user)
    db.commit()

# Felhasználó státusz módosítása (tiltás/engedélyezés)
@router.patch("/users/{user_id}/status", status_code=204)
def update_user_status(
    user_id: int,
    payload: StatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_moderator)
):
    user = db.get(User, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="Felhasználó nem található")
    
    # Ne lehessen saját magad tiltani
    if user.id == current_user.id:
        raise HTTPException(status_code=400, detail="Nem tilthatod le saját magad")
    
    # Státusz módosítása: -1 = tiltott, 0+ = aktív
    if payload.disabled:
        user.role_id = -1  # Tiltott státusz
    else:
        user.role_id = 0  # Visszaállítás user szerepkörre
    
    db.add(user)
    db.commit()

# Felhasználó törlése
@router.delete("/users/{user_id}", status_code=204)
def delete_user(
    user_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_moderator)
):
    # Csak admin törölhet felhasználót
    if current_user.role_id < 2:
        raise HTTPException(status_code=403, detail="Csak admin törölhet felhasználót")
    
    user = db.get(User, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="Felhasználó nem található")
    
    # Ne lehessen saját magad törölni
    if user.id == current_user.id:
        raise HTTPException(status_code=400, detail="Nem törölheted saját magad")
    
    db.delete(user)
    db.commit()

File: frontend/src/lib/api/moderator.ts
import { apiClient } from './client';

export interface UserListItem {
  id: number;
  username: string;
  email: string;
  full_name?: string;
  phone?: string;
  role_id: number;
  created_at: string;
  updated_at: string;
}

export interface UserDetails extends UserListItem {
  preferences: string[];
  sensitivities: string[];
}

const API_BASE = 'http://127.0.0.1:8000';

// Felhasználók listázása (moderátor)
export async function getAllUsers(): Promise<UserListItem[]> {
  const res = await apiClient.get(`${API_BASE}/api/v1/moderator/users`);
  return res.data;
}

// Felhasználó részletes adatai
export async function getUserDetails(userId: number): Promise<UserDetails> {
  const [userRes, prefsRes, sensRes] = await Promise.all([
    apiClient.get(`${API_BASE}/api/v1/users/${userId}`),
    apiClient.get(`${API_BASE}/api/v1/users/${userId}/preferences/`),
    apiClient.get(`${API_BASE}/api/v1/users/${userId}/sensitivities/`),
  ]);
  
  return {
    ...userRes.data,
    preferences: prefsRes.data,
    sensitivities: sensRes.data,
  };
}

// Szerepkör módosítása
export async function updateUserRole(userId: number, roleId: number): Promise<void> {
  await apiClient.patch(`${API_BASE}/api/v1/moderator/users/${userId}/role`, { role_id: roleId });
}

// Felhasználó tiltása/engedélyezése
export async function toggleUserStatus(userId: number, disabled: boolean): Promise<void> {
  await apiClient.patch(`${API_BASE}/api/v1/moderator/users/${userId}/status`, { disabled });
}

// Felhasználó törlése
export async function deleteUser(userId: number): Promise<void> {
  await apiClient.delete(`${API_BASE}/api/v1/moderator/users/${userId}`);
}
